### PR TITLE
 Add new Kafka Role to the docs

### PR DESCRIPTION
**Pull request by manalibhutiyani**
Fixes: #3118
    Signed-off-by: Manali Bhutiyani <manali@covalent.io>
```release-note
Add new Kafka Role to the docs
```

### CODE REVIEW

**Comment by manalibhutiyani** [comment]
test-me-please

**Comment by manalibhutiyani** [comment]
test-me-please

**Comment by manalibhutiyani** [comment]
test-me-please

File: Documentation/policy/language.rst [hunk 1]
```diff
@@ -527,18 +527,29 @@ if all fields are empty or missing, the rule will match all Kafka messages.
 
 The following fields can be matched on:
 
-
-APIVersion
-  APIVersion is the version matched against the api version of the Kafka
-  message. If set, it has to be a string representing a positive integer. If
-  omitted or empty, all versions are allowed.
+Role
+  Role is a case-insensitive string and describes a group of API keys
```
Comment (joestringer) [inline Documentation/policy/language.rst]: case-insensitive string *which*  describes...

File: Documentation/policy/language.rst [hunk 2]
```diff
@@ -527,18 +527,29 @@ if all fields are empty or missing, the rule will match all Kafka messages.
...
+  necessary to perform certain higher level Kafka operations such as "produce"
+  or "consume". A Role automatically expands into all APIKeys required
+  to perform the specified higher level operation.
+
+  The following values are supported:
+	- "produce": Allow producing to the topics specified in the rule
+	- "consume": Allow consuming from the topics specified in the rule
+  This field is incompatible with the APIKey field, i.e either APIKey or Role
```
Comment (joestringer) [inline Documentation/policy/language.rst]: `either APIKey or Role can be specified` -> `APIKey and Role cannot both be specified in the same rule`

File: pkg/policy/api/rule.go [hunk 3]
```diff
@@ -438,19 +438,18 @@ type PortRuleHTTP struct {
...
 	// The following values are supported:
 	//  - "produce": Allow producing to the topics specified in the rule
 	//  - "consume": Allow consuming from the topics specified in the rule
 	//
-	// This field is incompatible with the APIKey field, either APIKey or Role
-	// may be specified.
-	//
-	// If omitted or empty, the field has no effect and the logic of the APIKey
-	// field applies.
+	// This field is incompatible with the APIKey field, i.e either APIKey or Role
```
Comment (joestringer) [inline pkg/policy/api/rule.go]: If you take my suggestion above, this one needs updating too.

**Review by joestringer — COMMENTED** [review]
Looks good, minor wording suggestions.

File: Documentation/policy/language.rst [hunk 4]
```diff
@@ -527,18 +527,30 @@ if all fields are empty or missing, the rule will match all Kafka messages.
...
+  necessary to perform certain higher level Kafka operations such as "produce"
+  or "consume". A Role automatically expands into all APIKeys required
+  to perform the specified higher level operation.
+
+  The following values are supported:
+	- "produce": Allow producing to the topics specified in the rule
+	- "consume": Allow consuming from the topics specified in the rule
+  This field is incompatible with the APIKey field, either APIKey or Role
+  may be specified. If omitted or empty, all keys are allowed, if APIKey is also
+  the empty
```
Comment (ianvernon) [inline Documentation/policy/language.rst]: ... is also empty

**Review by ianvernon — CHANGES_REQUESTED** [review]

File: Documentation/policy/language.rst (hunk 1 above)
Comment (manalibhutiyani) [reply Documentation/policy/language.rst]: done

File: Documentation/policy/language.rst (hunk 2 above)
Comment (manalibhutiyani) [reply Documentation/policy/language.rst]: done

File: pkg/policy/api/rule.go (hunk 3 above)
Comment (manalibhutiyani) [reply pkg/policy/api/rule.go]: done

File: Documentation/policy/language.rst (hunk 4 above)
Comment (manalibhutiyani) [reply Documentation/policy/language.rst]: done.

**Comment by manalibhutiyani** [comment]
test-me-please

**Comment by manalibhutiyani** [comment]
test-me-please

**Comment by manalibhutiyani** [comment]
test-me-please

**Comment by manalibhutiyani** [comment]
@ianvernon @joestringer : Addressed review comments. Need an approval if everything looks ok.

File: Documentation/policy/language.rst [hunk 5]
```diff
@@ -527,18 +527,31 @@ if all fields are empty or missing, the rule will match all Kafka messages.
...
-APIVersion
-  APIVersion is the version matched against the api version of the Kafka
-  message. If set, it has to be a string representing a positive integer. If
-  omitted or empty, all versions are allowed.
+    - "produce": Allow producing to the topics specified in the rule.
+    - "consume": Allow consuming from the topics specified in the rule.
+
+  This field is incompatible with the APIKey field, i.e APIKey and Role
+  cannot both be specified in the same rule.
+  If omitted or empty, all keys are allowed.
```
Comment (tgraf) [inline Documentation/policy/language.rst]: This last sentence is not true. If omitted, APIKey can still be specified. This needs to be clear.

**Comment by manalibhutiyani** [comment]
@eloycoto : Still seeing this test flake unrelated to my docs changes

```jsonpath='{.items[*].metadata.deletionTimestamp}'"
21:46:33 
21:46:33 • Failure [200.479 seconds]
21:46:33 K8sValidatedPolicyTestAcrossNamespaces
21:46:33 /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/Policies.go:675
21:46:33   Policies Across Namespaces [It]
21:46:33   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/Policies.go:712
21:46:33 
21:46:33   Unable to connect between frontend-d9q97 and 10.97.83.36:80/
21:46:33   Expected
21:46:33       <string>: 000command terminated with exit code 7
21:46:33       
21:46:33   to equal
21:46:33       <string>: 200
21:46:33 
21:46:33   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/Policies.go:855
21:46:33 ------------------------------
21:46:33 SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS```

**Comment by manalibhutiyani** [comment]
test-me-please

**Comment by manalibhutiyani** [comment]
And this one: 
```
jsonpath='{.items[*].metadata.deletionTimestamp}'"
21:46:33 
21:46:33 • Failure [200.479 seconds]
21:46:33 K8sValidatedPolicyTestAcrossNamespaces
21:46:33 /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/Policies.go:675
21:46:33   Policies Across Namespaces [It]
21:46:33   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/Policies.go:712
21:46:33 
21:46:33   Unable to connect between frontend-d9q97 and 10.97.83.36:80/
21:46:33   Expected
21:46:33       <string>: 000command terminated with exit code 7
21:46:33       
21:46:33   to equal
21:46:33       <string>: 200
21:46:33 
21:46:33   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/Policies.go:855
21:46:33 ------------------------------
21:46:33 SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS```

File: Documentation/policy/language.rst [hunk 6]
```diff
@@ -566,8 +579,25 @@ Topic
 
   If omitted or empty, all topics are allowed.
 
-Only allow producing to topic empire-announce
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Only allow producing to topic empire-announce using Role
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
```
Comment (tgraf) [inline Documentation/policy/language.rst:590]: This needs a sentence here that describes the motivation of role and when to use roles and when to use apiKeys.

File: examples/policies/l7/kafka/kafka-Role.yaml [hunk 7]
```diff
@@ -0,0 +1,23 @@
+apiVersion: "cilium.io/v2"
```
Comment (tgraf) [inline examples/policies/l7/kafka/kafka-Role.yaml:1]: Rename the filename to all lowercase

File: pkg/policy/api/rule.go [hunk 8]
```diff
@@ -438,19 +438,18 @@ type PortRuleHTTP struct {
...
 	//
-	// This field is incompatible with the APIKey field, either APIKey or Role
-	// may be specified.
-	//
-	// If omitted or empty, the field has no effect and the logic of the APIKey
-	// field applies.
+	// This field is incompatible with the APIKey field, i.e APIKey and Role
+	// cannot both be specified in the same rule.
 	//
+	// If omitted or empty, all keys are allowed.
```
Comment (tgraf) [inline pkg/policy/api/rule.go]: This needs to be adjusted.

**Review by tgraf — CHANGES_REQUESTED** [review]

File: Documentation/policy/language.rst (hunk 5 above)
Comment (manalibhutiyani) [reply Documentation/policy/language.rst]: Changing this to : `  If omitted or empty, and if APIKey is not specified then all keys are allowed.`

File: Documentation/policy/language.rst (hunk 6 above)
Comment (manalibhutiyani) [reply Documentation/policy/language.rst:590]: Planning to add this:
```
PortRuleKafka is a list of Kafka protocol constraints. All fields are optional,
if all fields are empty or missing, the rule will match all Kafka messages.
There are two ways to specify the Kafka rules. We can choose to specify a
high level "produce" or "consume" role to a topic or choose to specify more
low level kafka protocol specific apiKeys. Thus the rule allows us to specify both
high level role constructs as well as low level Kafka protocol constructs.
```

File: examples/policies/l7/kafka/kafka-Role.yaml (hunk 7 above)
Comment (manalibhutiyani) [reply examples/policies/l7/kafka/kafka-Role.yaml:1]: done

File: pkg/policy/api/rule.go (hunk 8 above)
Comment (manalibhutiyani) [reply pkg/policy/api/rule.go]: Changed to `If omitted or empty,
  and if Role is not specified, then all keys are allowed.`

**Review by tgraf — APPROVED** [review]

File: Documentation/policy/language.rst (hunk 6 above)
Comment (manalibhutiyani) [reply Documentation/policy/language.rst:590]: done, please let me know if you think this needs to be improved.

**Comment by manalibhutiyani** [comment]
test-me-please

**Comment by manalibhutiyani** [comment]
I am seeing a lot of unrelated CI errors. Not sure what's the best route here? Trying CI again.

**Comment by manalibhutiyani** [comment]
test-me-please

**Comment by tgraf** [comment]
> I am seeing a lot of unrelated CI errors. Not sure what's the best route here? Trying CI again.

All PRs are seeing these right now. Nee to investigate tomorrow.

File: Documentation/policy/language.rst [hunk 9]
```diff
@@ -524,20 +524,38 @@ Kafka (Tech Preview)
 
 PortRuleKafka is a list of Kafka protocol constraints. All fields are optional,
 if all fields are empty or missing, the rule will match all Kafka messages.
+There are two ways to specify the Kafka rules. We can choose to specify a
+high level "produce" or "consume" role to a topic or choose to specify more
```
Comment (ianvernon) [inline Documentation/policy/language.rst]: high-level

File: Documentation/policy/language.rst [hunk 10]
```diff
@@ -524,20 +524,38 @@ Kafka (Tech Preview)
...
+low level kafka protocol specific apiKeys. Thus the rule allows us to specify
```
Comment (ianvernon) [inline Documentation/policy/language.rst]: low-level

kafka --> Kafka

File: Documentation/policy/language.rst [hunk 11]
```diff
@@ -524,20 +524,38 @@ Kafka (Tech Preview)
...
+both high level role constructs as well as low level Kafka protocol constructs.
```
Comment (ianvernon) [inline Documentation/policy/language.rst]: high-level ... low-level

File: Documentation/policy/language.rst [hunk 12]
```diff
@@ -524,20 +524,38 @@ Kafka (Tech Preview)
...
 
 The following fields can be matched on:
 
+Role
+  Role is a case-insensitive string which describes a group of API keys
+  necessary to perform certain higher level Kafka operations such as "produce"
```
Comment (ianvernon) [inline Documentation/policy/language.rst]: higher-level

File: Documentation/policy/language.rst [hunk 13]
```diff
@@ -524,20 +524,38 @@ Kafka (Tech Preview)
...
+  or "consume". A Role automatically expands into all APIKeys required
+  to perform the specified higher level operation.
```
Comment (ianvernon) [inline Documentation/policy/language.rst]: higher-level

**Review by ianvernon — CHANGES_REQUESTED** [review]

File: Documentation/policy/language.rst (hunk 11 above)
Comment (joestringer) [reply Documentation/policy/language.rst]: These last two sentences appear to state the same thing. Perhaps replace the last sentence with something like:

"Writing rules based on Kafka roles is easier and covers most common use cases, however if more granularity is needed then users can alternatively write rules using specific apiKeys."

Comment (manalibhutiyani) [reply Documentation/policy/language.rst]: sure.

File: Documentation/policy/language.rst (hunk 10 above)
Comment (manalibhutiyani) [reply Documentation/policy/language.rst]: done

File: Documentation/policy/language.rst (hunk 11 above)
Comment (manalibhutiyani) [reply Documentation/policy/language.rst]: done

File: Documentation/policy/language.rst (hunk 12 above)
Comment (manalibhutiyani) [reply Documentation/policy/language.rst]: done

File: Documentation/policy/language.rst (hunk 13 above)
Comment (manalibhutiyani) [reply Documentation/policy/language.rst]: done

File: Documentation/policy/language.rst (hunk 9 above)
Comment (manalibhutiyani) [reply Documentation/policy/language.rst]: done

**Comment by manalibhutiyani** [comment]
test-me-please

**Pull request merged**